metricbeat/helper/metricset.go

```diff
@@ -1,195 +1,87 @@
 package helper
 
 import (
-	"github.com/elastic/beats/libbeat/beat"
-	"github.com/elastic/beats/libbeat/common"
-	"github.com/elastic/beats/libbeat/logp"
-	"gopkg.in/yaml.v2"
-	"sync"
 	"time"
-)
 
-// Base metric configuration
-type MetricSetConfig struct {
-	Period string
-}
+	"github.com/elastic/beats/libbeat/common"
+)
 
 // Metric specific data
 // This must be defined by each metric
 type MetricSet struct {
-	Name    string
-	Enabled bool
-
-	// Generic Config existing in all metrics
-	BaseConfig MetricSetConfig
-
-	// Raw metric specific config
-	// This is provided to convert it into Config later
-	RawConfig interface{}
-
-	// Metric specific config
-	Config interface{}
-
+	Name        string
 	MetricSeter MetricSeter
-	Module      *Module
-
-	// Control channel
-	done chan struct{}
-}
-
-// Interface for each metric
-type MetricSeter interface {
-	// Setup needed for all upcoming fetches
-	// Typically config is loaded here
-	Setup() error
-
-	// Method to periodically fetch new events
-	Fetch() ([]common.MapStr, error)
-
-	// Cleanup when stopping metricset
-	Cleanup() error
+	// Inherits config from module
+	Config ModuleConfig
 }
 
 // Creates a new MetricSet
-func NewMetricSet(name string, metricset MetricSeter, module *Module) *MetricSet {
+func NewMetricSet(name string, metricset MetricSeter, config ModuleConfig) *MetricSet {
 	return &MetricSet{
 		Name:        name,
 		MetricSeter: metricset,
-		Module:      module,
-		Enabled:     false,
-		done:        make(chan struct{}),
+		Config:      config,
 	}
 }
 
-func (m *MetricSet) LoadConfig(config interface{}) {
+// RunMetric runs the given metricSet and returns the event
+func (m *MetricSet) Fetch() ([]common.MapStr, error) {
 
-	bytes, err := yaml.Marshal(m.RawConfig)
+	events, err := m.MetricSeter.Fetch(m)
 
 	if err != nil {
-		logp.Err("Load metric config error: %v", err)
+		return nil, err
 	}
-	yaml.Unmarshal(bytes, config)
-}
+	newEvents := []common.MapStr{}
 
-// Registers metric with module
-func (m *MetricSet) Register() {
-	m.Module.AddMetric(m)
-}
+	// Default names based on module and metric
+	// These can be overwritten by setting index or / and type in the event
+	indexName := ""
 
-// RunMetric runs the given metric
-func (m *MetricSet) Start(b *beat.Beat, wg sync.WaitGroup) {
+	// Type is the same for all metricsets
+	typeName := "metricsets"
+	timestamp := common.Time(time.Now())
 
-	// Catches metric in case of panic. Keeps other metricsets running
-	defer func() {
-		if r := recover(); r != nil {
-			logp.Err("Metric %s paniced and stopped running. Reason: %+v", m.Name, r)
+	for _, event := range events {
+		// Set index from event if set
+		if _, ok := event["index"]; ok {
+			indexName = event["index"].(string)
```

Comment:
In the future this should be solved through https://github.com/elastic/beats/issues/923